File: frontend/src/lib/api/admin_resources.ts
/**
 * Typed API client for the admin resource CRUD surface (T26.5).
 *
 * Six endpoints under ``/api/admin/resources``, all gated by
 * ``require_role("admin")`` on the backend (S22 cookie pattern). Mirrors
 * the S25 :file:`cities_admin.ts` shape post-/reviewing-and-fixing
 * lift: every function delegates transport to :func:`fetchWithCookie`
 * (which bakes in ``credentials: "include"``) and error-shaping to
 * :func:`throwOnApiError` from `_client.ts` — one typed error class
 * per domain, never a per-domain copy of the timeout/cookie wrapper.
 *
 * Wire types match the backend dict shape returned by
 * :mod:`app.core.queries_admin_resources` 1:1, including the
 * server-stamped ``user_curated_at`` (T26.1's loader-respect contract:
 * any row with this column populated survives the seed loader).
 */

import { fetchWithCookie, throwOnApiError } from "./_client";

/* -------------------------------------------------------------------------
 * Wire types — match the backend row dict 1:1.
 * ------------------------------------------------------------------------- */

/**
 * One row from the ``resources`` table as exposed to admin consumers.
 *
 * ``health_status`` is one of ``'healthy' | 'watch' | 'flagged' |
 * 'hidden'`` — kept as a plain string here so the UI can render
 * future statuses without a client redeploy. ``user_curated_at`` is
 * server-stamped (ISO-8601 UTC) on every write; ``null`` only for
 * legacy rows that predate the admin surface.
 */
export interface Resource {
  id: number;
  name: string;
  category: string;
  subcategory: string | null;
  address: string | null;
  lat: number | null;
  lng: number | null;
  phone: string | null;
  url: string | null;
  eligibility: string | null;
  services: string | null;
  hours: string | null;
  notes: string | null;
  city: string;
  barrier_affinity: string | null;
  health_status: string;
  user_curated_at: string | null;
}

/**
 * Paginated list response. ``limit`` + ``offset`` are echoed so the UI
 * can render "showing N–M of total" without re-deriving them.
 */
export interface ResourceListResponse {
  items: Resource[];
  total: number;
  limit: number;
  offset: number;
}

/**
 * POST body for :func:`createResource`. Mirrors the backend
 * ``CreateResourceBody`` Pydantic model — required: ``name``,
 * ``category``, ``city``; optional fields cover the writable subset.
 * ``health_status`` is forced server-side; ``user_curated_at`` is
 * stamped server-side; neither is client-supplied.
 */
export interface CreateResourcePayload {
  name: string;
  category: string;
  city: string;
  subcategory?: string | null;
  address?: string | null;
  lat?: number | null;
  lng?: number | null;
  phone?: string | null;
  url?: string | null;
  eligibility?: string | null;
  services?: string | null;
  hours?: string | null;
  notes?: string | null;
  barrier_affinity?: string | null;
}

/**
 * PATCH body for :func:`updateResource` — partial update; every field
 * optional. The backend stamps ``user_curated_at`` even when no other
 * field changes (touch-as-curation semantics).
 */
export interface UpdateResourcePatch {
  name?: string;
  category?: string;
  city?: string;
  subcategory?: string | null;
  address?: string | null;
  lat?: number | null;
  lng?: number | null;
  phone?: string | null;
  url?: string | null;
  eligibility?: string | null;
  services?: string | null;
  hours?: string | null;
  notes?: string | null;
  barrier_affinity?: string | null;
}

/**
 * Response from :func:`restoreResource` — the backend echoes the
 * resulting state rather than re-fetching the whole row.
 */
export interface RestoreResourceResponse {
  id: number;
  health_status: string;
}

/* -------------------------------------------------------------------------
 * Error class — single typed discriminator for callers (status code).
 * ------------------------------------------------------------------------- */

export class AdminResourcesApiError extends Error {
  readonly status: number;
  readonly detail?: string;

  constructor(status: number, message: string, detail?: string) {
    super(message);
    this.name = "AdminResourcesApiError";
    this.status = status;
    this.detail = detail;
  }
}

/* -------------------------------------------------------------------------
 * Internal helpers.
 * ------------------------------------------------------------------------- */

const BASE = "/api/admin/resources";

interface ListResourcesOpts {
  city?: string;
  limit?: number;
  offset?: number;
  includeHidden?: boolean;
}

/**
 * Build the ``?city=...&limit=...`` query string for :func:`listResources`.
 *
 * Returns ``""`` (no leading ``?``) when no opts are present so the URL
 * stays clean. ``includeHidden`` maps to the snake_case backend param
 * ``include_hidden`` — translation kept here so callers think in JS
 * conventions.
 */
function _buildListQuery(opts: ListResourcesOpts): string {
  const params = new URLSearchParams();
  if (opts.city !== undefined) params.set("city", opts.city);
  if (opts.limit !== undefined) params.set("limit", String(opts.limit));
  if (opts.offset !== undefined) params.set("offset", String(opts.offset));
  if (opts.includeHidden !== undefined) {
    params.set("include_hidden", String(opts.includeHidden));
  }
  const qs = params.toString();
  return qs ? `?${qs}` : "";
}

/* -------------------------------------------------------------------------
 * Public surface — one function per backend route.
 * ------------------------------------------------------------------------- */

export async function listResources(
  opts: ListResourcesOpts = {},
): Promise<ResourceListResponse> {
  const res = await fetchWithCookie(`${BASE}${_buildListQuery(opts)}`, {
    method: "GET",
  });
  await throwOnApiError(res, AdminResourcesApiError, "admin-resources");
  return (await res.json()) as ResourceListResponse;
}

export async function getResource(id: number): Promise<Resource> {
  const res = await fetchWithCookie(`${BASE}/${id}`, { method: "GET" });
  await throwOnApiError(res, AdminResourcesApiError, "admin-resources");
  return (await res.json()) as Resource;
}

export async function createResource(
  payload: CreateResourcePayload,
): Promise<Resource> {
  const res = await fetchWithCookie(BASE, {
    method: "POST",
    body: JSON.stringify(payload),
  });
  await throwOnApiError(res, AdminResourcesApiError, "admin-resources");
  return (await res.json()) as Resource;
}

export async function updateResource(
  id: number,
  patch: UpdateResourcePatch,
): Promise<Resource> {
  const res = await fetchWithCookie(`${BASE}/${id}`, {
    method: "PATCH",
    body: JSON.stringify(patch),
  });
  await throwOnApiError(res, AdminResourcesApiError, "admin-resources");
  return (await res.json()) as Resource;
}

export async function hideResource(id: number): Promise<void> {
  const res = await fetchWithCookie(`${BASE}/${id}`, { method: "DELETE" });
  await throwOnApiError(res, AdminResourcesApiError, "admin-resources");
}

export async function restoreResource(
  id: number,
): Promise<RestoreResourceResponse> {
  const res = await fetchWithCookie(`${BASE}/${id}/restore`, {
    method: "POST",
  });
  await throwOnApiError(res, AdminResourcesApiError, "admin-resources");
  return (await res.json()) as RestoreResourceResponse;
}

File: backend/app/routes/admin_resources.py
"""Admin resource CRUD routes (T26.2).

Six endpoints under ``/api/admin/resources/...``, all gated by
:func:`require_role("admin")` (S22 cookie pattern):

* ``GET    /api/admin/resources?city=&limit=&offset=&include_hidden=``
* ``GET    /api/admin/resources/{id}``
* ``POST   /api/admin/resources`` (returns ``{id, ...}``)
* ``PATCH  /api/admin/resources/{id}``
* ``DELETE /api/admin/resources/{id}`` (soft-hide via set_health_status)
* ``POST   /api/admin/resources/{id}/restore`` (sets health_status='healthy')

Every write path stamps ``resources.user_curated_at = now()``, consuming
T26.1's loader-respect contract: rows the admin curates here are not
overwritten by the seed loader on container restart.

This module is *additive*: it ships in a new file rather than extending
``routes/admin_feedback.py`` because the feedback surface is scoped to
flagged-queue + visit-inbox actions; mixing CRUD into that file would
muddy the trust+audit story (different audit table contracts).
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query, Response
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.core import queries_admin_resources
from app.core.auth_roles import require_role
from app.core.database import get_db


router = APIRouter(prefix="/api/admin/resources", tags=["admin", "resources"])

_require_admin = require_role("admin")


class CreateResourceBody(BaseModel):
    """POST /api/admin/resources body.

    Required: ``name``, ``category``, ``city``. Optional fields mirror
    the writable subset of ``resources`` columns; ``health_status`` is
    forced to ``'healthy'`` server-side and ``user_curated_at`` is
    stamped server-side — neither is client-supplied.
    """

    name: str = Field(min_length=1, max_length=200)
    category: str = Field(min_length=1, max_length=80)
    city: str = Field(min_length=1, max_length=80)
    subcategory: str | None = None
    address: str | None = None
    lat: float | None = None
    lng: float | None = None
    phone: str | None = None
    url: str | None = None
    eligibility: str | None = None
    services: str | None = None
    hours: str | None = None
    notes: str | None = None
    barrier_affinity: str | None = None


class UpdateResourcePatch(BaseModel):
    """PATCH /api/admin/resources/{id} body — partial update.

    Every field is optional; the route stamps ``user_curated_at`` even
    when no other field changed (touch-as-curation semantics).
    """

    name: str | None = Field(default=None, min_length=1, max_length=200)
    category: str | None = Field(default=None, min_length=1, max_length=80)
    city: str | None = Field(default=None, min_length=1, max_length=80)
    subcategory: str | None = None
    address: str | None = None
    lat: float | None = None
    lng: float | None = None
    phone: str | None = None
    url: str | None = None
    eligibility: str | None = None
    services: str | None = None
    hours: str | None = None
    notes: str | None = None
    barrier_affinity: str | None = None


@router.get("")
async def list_resources(
    city: str | None = Query(default=None),
    limit: int = Query(default=50, ge=1, le=100),
    offset: int = Query(default=0, ge=0),
    include_hidden: bool = Query(default=False),
    db: AsyncSession = Depends(get_db),
    _account: dict = Depends(_require_admin),
) -> dict:
    """Paginated browse of resources.

    Default page size 50; FastAPI clamps ``limit`` at 100 (returns 422
    above that). ``include_hidden=true`` opts soft-deleted rows back in;
    ``city=<slug>`` scopes to one city. Response echoes ``limit`` +
    ``offset`` + ``total`` so the UI can render "showing N–M of total".
    """
    rows, total = await queries_admin_resources.list_resources(
        db, city=city, limit=limit, offset=offset,
        include_hidden=include_hidden,
    )
    return {
        "items": rows,
        "total": total,
        "limit": limit,
        "offset": offset,
    }


@router.get("/{resource_id}")
async def get_resource(
    resource_id: int,
    db: AsyncSession = Depends(get_db),
    _account: dict = Depends(_require_admin),
) -> dict:
    """Return a single resource by id, or 404 when absent."""
    row = await queries_admin_resources.get_resource(db, resource_id)
    if row is None:
        raise HTTPException(status_code=404, detail="Resource not found")
    return row


@router.post("", status_code=201)
async def create_resource(
    body: CreateResourceBody,
    db: AsyncSession = Depends(get_db),
    _account: dict = Depends(_require_admin),
) -> dict:
    """Create a new resource. Stamps ``user_curated_at = now()``.

    Returns the full created row (including the new ``id``) so the UI
    can update its table without a follow-up GET.
    """
    fields = body.model_dump(exclude_none=True)
    new_id = await queries_admin_resources.create_resource(db, **fields)
    created = await queries_admin_resources.get_resource(db, new_id)
    if created is None:
        # Defensive: created in this transaction, must exist on read-back.
        raise HTTPException(
            status_code=500, detail="Resource not retrievable after create"
        )
    return created


@router.patch("/{resource_id}")
async def update_resource(
    resource_id: int,
    patch: UpdateResourcePatch,
    db: AsyncSession = Depends(get_db),
    _account: dict = Depends(_require_admin),
) -> dict:
    """Patch a resource. Stamps ``user_curated_at = now()`` on every call.

    404 when *resource_id* does not exist. Returns the full updated row.
    """
    fields = patch.model_dump(exclude_none=True)
    matched = await queries_admin_resources.update_resource(
        db, resource_id, **fields
    )
    if not matched:
        raise HTTPException(status_code=404, detail="Resource not found")
    updated = await queries_admin_resources.get_resource(db, resource_id)
    if updated is None:
        raise HTTPException(
            status_code=500, detail="Resource not retrievable after update"
        )
    return updated


@router.delete("/{resource_id}", status_code=204)
async def hide_resource(
    resource_id: int,
    db: AsyncSession = Depends(get_db),
    _account: dict = Depends(_require_admin),
) -> Response:
    """Soft-delete: set ``health_status='hidden'``.

    The row stays in the table so audit trails (feedback rows that
    reference it) remain intact. Use POST /restore to reverse. 204 on
    success; 404 when *resource_id* does not exist.
    """
    matched = await queries_admin_resources.set_health_status(
        db, resource_id, "hidden"
    )
    if not matched:
        raise HTTPException(status_code=404, detail="Resource not found")
    return Response(status_code=204)


@router.post("/{resource_id}/restore")
async def restore_resource(
    resource_id: int,
    db: AsyncSession = Depends(get_db),
    _account: dict = Depends(_require_admin),
) -> dict:
    """Reverse a soft-delete: set ``health_status='healthy'``.

    Returns the new ``health_status`` for echoing back to the UI.
    404 when *resource_id* does not exist.
    """
    matched = await queries_admin_resources.set_health_status(
        db, resource_id, "healthy"
    )
    if not matched:
        raise HTTPException(status_code=404, detail="Resource not found")
    return {"id": resource_id, "health_status": "healthy"}


__all__ = ["router", "CreateResourceBody", "UpdateResourcePatch"]
